docs/rfcs/033-ancestor-deletion.md

```diff
@@ -0,0 +1,237 @@
+                                                         |
+                                                         |
+                                                         |
+                                                         -> child B
+
+|----"new main" timeline--------->
+                 |
+                 |-> child A
+
+
+```
+
+### After removing timelines
+
+We end up with a totally clean state that leaves no trace that a rollback
+ever happened: there is only one root timeline.
+
+```
+| ----"new main" timeline----------->
+                |
+                |-> child A
+
+
+```
+
+## Caveats
+
+Important things for API users to bear in mind:
+
+- this API does not delete the parent timeline: you must still do that explicitly.
+- if there are other child timelines ahead of the branch point of the merged
+  child, the parent won't be deletable: you must either delete or merge those
+  children.
+- do _not_ simply loop over all children and merge them all: this can have an
+  extremely high storage cost. The merge API is intended for use on a single
+  timeline to make it the new "main".
+- The merge API should also not be
+  exposed directly to the user as button/API, because they might decide
+  to click it for all the children and thereby generate many copies of the
+  parent's data -- the merge API should be used as part
+  of a high level "clean up after rollback" feature.
+
+## `merge_parent` API implementation
+
+Terms used in the following sections:
+
+- "the child": the timeline whose ID is specified in the merge API URL, also
+  called "new main" in the example.
+- "the parent": the parent of "the child". Also called "old main" in the example.
+- "the branch point" the ancestor_lsn of "the child"
+
+### Phase 1: write out adopted layers to S3
+
+The child will "adopt" layers from the parent, such that its end state contains
+all the parent's history as well as its own.
+
+For all layers in the parent's layer map whose high LSN is below the branch
+point, issue S3 CopyObject requests to duplicate them into the child timeline's
+prefix. Do not add them to the child's layer map yet.
+
+For delta layers in the parent's layer map which straddle the branch point, read them
+and write out only content up to the branch point into new layer objects.
+
+This is a long running operation if the parent has many layers: it should be
+implemented in a way that resumes rather than restarting from scratch, if the API
+times out and is called again.
+
+As an optimization, if there are no other timelines that will be adopted into
+the child, _and_ the child's image layers already full cover the branch LSN,
+then we may skip adopting layers.
+
+### Phase 2: update the child's index
+
+Having written out all needed layers in phase 1, atomically link them all
+into the child's IndexPart and upload to S3. This may be done while the
+child Timeline is still running.
```

Comment:
This does leave us in a pickle:

We would continue to deny requests to delete the "old main" timeline while these would exist.

At runtime we don't have a way to change ancestor right now, but I cannot see why not an sync RwLock could be used there (`Timeline::ancestor_timeline`)... An RwLock would complicate all of the operations, and for example cancel/shutdown propagation from ancestor `Timeline::get` to reparented child `Timeline::get`, but that situation is pathological as well.

Interestingly the `Timeline::get_ancestor_timeline_id` is more used than `Timeline::ancestor_timeline` or `Timeline::get_ancestor_timeline`. Well, after a quick clance I dont think there are any major issues.

Reply:
Attempted the above idea in #7285.